lib/service/type.rb

```diff
@@ -66,6 +66,19 @@ def all
           end
       end
 
+      # The enabled list of services may contain references to unknown types
+      # This can happen when a service is uninstalled without being disabled first
+      # In these cases, the service should be skipped
+      def safe_enabled_types
+        @enabled_types ||= enabled.map do |id|
+          begin
+            self[id]
+          rescue UnknownServiceTypeError
+            $stderr.puts "Skipping unknown service: #{id}"
+          end
+        end.reject(&:nil?)
+      end
```

Comment:
Good catch with this issue.

The naming of this method and `enabled` could do with some work.  I was initially confused that this was returning an array of a different class of objects to `enabled`.  A better naming would be:

* Rename `Type.enabled` to `Type.enabled_names`
* Similar with `Type#enabled`
* Rename `Type.safe_enabled_types` to `Type.enabled`.

Reply:
I have taken a different approach because:

* `enabled_names` is a misnomer as missing `types` are not enabled (they are missing),
* It isn't immediately clearer if a `enabled` or `enabled_names` should be used, and
* I dislike how `enabled_names` opens the eigenclass to patch in a `save` method.

Instead the methods have been refactored into an `enabled` and `save_enabled_file` methods. This means:
* `Type.enabled` returns a regular array of enabled `Type` objects, and
* `Type.save_enabled_file` will save the current state of the previous method.

NOTE: This does mean "missing enabled types" will be implicitly disabled the next time the file is saved. This is because they do not appear in  the `enabled` list. This should not be an issue, as the missing `type` has probably been uninstalled.

Reply:
50250a1dd66f49543eaf0d209f9690f46b0d3740

Reply:
This has introduced a change in behaviour.  Consider the following:

```
sudo yum install flight-login-api
flight service enable login-api
sudo yum remove flight-login-api
flight service disable www; flight service enable www # one of these will have caused a change.
sudo yum install flight-login-api
```

Previously, following the reinstallation of `flight-login-api` it would have been enabled.  With this change, it is now disabled.

This is at odds with how `systemctl`/`systemd` does things.  There uninstalling services does not change their enabled status.

Perhaps, `save_enabled_file` could call `enabled(include_missing: true)`?

Reply:
I agree that this is a change in behaviour but I question if it matters because:
* The previous behaviour was to break if a type was missing,
* We don't have parity with `systemd` in multiple other ways, and
* Keeping them "pseudo-enabled" means they should be a valid input to `flight service disable`

Making them a valid input to `flight service disable` is probably more effort than it is worth as `Type['missing']` raises an error. It also means `Type#disable` can not be used to disable these "pseudo-enabled" types.

Something to consider before we add complexity to a rather esoteric use case.

Reply:
We seem to be talking at cross purposes here.  Let me try again.

I want the following interaction to *not* change.  That is, the enabled/disabled status is remembered on reinstall.

```
[flight@gateway1 (ben1) ~]$ flight service avail | grep console-api | cut -f 4
enabled
[flight@gateway1 (ben1) ~]$ sudo yum -y remove flight-console-api > /dev/null
[flight@gateway1 (ben1) ~]$ ( flight service enable www ; flight service disable www ) > /dev/null
[flight@gateway1 (ben1) ~]$ sudo yum -y install flight-console-api > /dev/null
[flight@gateway1 (ben1) ~]$ flight service avail | grep console-api | cut -f 4
enabled
```

I also want the following interaction to *not* change.  That is, attempting to enable or disable a uninstalled (or simply mis-typed) service results in an error.

```
[flight@gateway1 (ben1) ~]$ flight service enable console-api
Service 'console-api' already enabled.
[flight@gateway1 (ben1) ~]$ sudo yum -y remove flight-console-api > /dev/null
[flight@gateway1 (ben1) ~]$ flight service enable console-api
flight service: unknown service type: console-api
```

I want the following interaction to change.

```
[flight@gateway1 (ben1) ~]$ flight service stack status
┌──────────────────┬────────┬───────┐
│ Service          │ Status │ PID   │
├──────────────────┼────────┼───────┤
│ desktop-restapi  │ active │ 20947 │
│ file-manager-api │ active │ 16482 │
│ job-script-api   │ active │ 21140 │
│ login-api        │ active │ 16423 │
│ console-api      │ active │ 19534 │
└──────────────────┴────────┴───────┘
[flight@gateway1 (ben1) ~]$ sudo yum -y remove flight-console-api > /dev/null
[flight@gateway1 (ben1) ~]$ flight service stack status
flight service: unknown service type: console-api
```

The implementation as of 90ce9a023e83ea36e745c9467befc6da58122b53 gets the first of these wrong.  It gets the second correct and fixes the issue in the third.

Reply:
https://github.com/openflighthpc/flight-service/pull/8/commits/cae9a2b10b9cfc7df653ca9a08142e89f42e1f60